pkg/ec2helper/ec2helper.go

```diff
@@ -1363,21 +1364,104 @@ func HasEbsVolume(image *ec2.Image) bool {
 	launchIdentifier := uuid.New()
 
 	fmt.Println("Creating Launch Template...")
-	input := &ec2.CreateLaunchTemplateInput{
-		LaunchTemplateData: &ec2.RequestLaunchTemplateData{
-			ImageId:      &simpleConfig.ImageId,
-			InstanceType: &simpleConfig.InstanceType,
-			NetworkInterfaces: []*ec2.LaunchTemplateInstanceNetworkInterfaceSpecificationRequest{
-				{
-					AssociatePublicIpAddress: aws.Bool(true),
-					DeviceIndex:              aws.Int64(0),
-				},
-			},
+
+	// Create and set network connections
+	networkInterfaces := []*ec2.LaunchTemplateInstanceNetworkInterfaceSpecificationRequest{
+		{
+			AssociatePublicIpAddress: aws.Bool(true),
+			DeviceIndex:              aws.Int64(0),
 		},
+	}
+	if simpleConfig.SecurityGroupIds != nil && len(simpleConfig.SecurityGroupIds) > 0 {
+		for _, network := range networkInterfaces {
+			network.SetGroups(aws.StringSlice(simpleConfig.SecurityGroupIds))
+		}
+	}
+	if simpleConfig.SubnetId != "" {
+		for _, network := range networkInterfaces {
+			network.SetSubnetId(simpleConfig.SubnetId)
+		}
+	}
+
+	// Define and set template data
+	templateData := &ec2.RequestLaunchTemplateData{
+		NetworkInterfaces: networkInterfaces,
+	}
+	if simpleConfig.IamInstanceProfile != "" {
+		templateData.SetIamInstanceProfile(&ec2.LaunchTemplateIamInstanceProfileSpecificationRequest{
+			Name: aws.String(simpleConfig.IamInstanceProfile),
+		})
+	}
+	if simpleConfig.ImageId != "" {
+		templateData.SetImageId(simpleConfig.ImageId)
+	}
+	if simpleConfig.InstanceType != "" {
+		templateData.SetInstanceType(simpleConfig.InstanceType)
+	}
+
+	setAutoTermination := false
+	if detailedConfig != nil {
+		// Set all EBS volumes not to be deleted, if specified
+		if HasEbsVolume(detailedConfig.Image) && simpleConfig.KeepEbsVolumeAfterTermination {
+			blockDevices := []*ec2.LaunchTemplateBlockDeviceMappingRequest{}
+			for index, block := range detailedConfig.Image.BlockDeviceMappings {
+				blockDevices = append(blockDevices, &ec2.LaunchTemplateBlockDeviceMappingRequest{
+					DeviceName:  block.DeviceName,
+					NoDevice:    block.NoDevice,
+					VirtualName: block.VirtualName,
+				})
+				if block.Ebs != nil {
+					blockDeviceEbs := &ec2.LaunchTemplateEbsBlockDeviceRequest{
+						DeleteOnTermination: aws.Bool(false),
+						Encrypted:           block.Ebs.Encrypted,
+						Iops:                block.Ebs.Iops,
+						KmsKeyId:            block.Ebs.KmsKeyId,
+						SnapshotId:          block.Ebs.SnapshotId,
+						Throughput:          block.Ebs.Throughput,
+						VolumeSize:          block.Ebs.VolumeSize,
+						VolumeType:          block.Ebs.VolumeType,
+					}
+					blockDevices[index].SetEbs(blockDeviceEbs)
+				}
+			}
+			templateData.SetBlockDeviceMappings(blockDevices)
+		}
+		setAutoTermination = IsLinux(*detailedConfig.Image.PlatformDetails) && simpleConfig.AutoTerminationTimerMinutes > 0
+	}
+
+	if setAutoTermination {
```

Comment:
Rather than copying and pasting [existing code](https://github.com/GavinBurris42/aws-simple-ec2-cli/blob/423e54630ec0d6aaa9a036b86d3fc4788f3046bd/pkg/ec2helper/ec2helper.go#L1041-L1064) for unique business logic like this, I'd rather we find a way to refactor the existing code so it can be re-used in both places. I know there are some subtle differences, so it will be a judgment call. But in this case, it should be possible to extract out the logic that constructs the UserData for the instance, even if actually setting the UserData on the request has to be in two places.